### ISSUE TITLE
Autogen based build fails due to multiple function definitions

### ISSUE THREAD

**Issue by marko-asplund** [JohnLangford/vowpal_wabbit#614]
Autogen based build fails due to multiple function definitions when linking libvw.so (current master branch [commit ID: a1fdc13]).

```
./autogen.sh
make
```

full typescript can be found here:
https://gist.github.com/anonymous/3efb4dd420ff308e42c9

here's an extract:

```
make
...
/bin/bash ../libtool  --tag=CXX   --mode=link g++ -I/usr/local/include -I/include  -Wall -Wno-unused-local-typedef    -pedan
tic  -ffast-math   -O3 -fomit-frame-pointer -fno-strict-aliasing -DNDEBUG  -std=gnu++11 -L/usr/local/lib -lboost_program_opt
ions -L/lib -lz -lpthread  -o libvw.la -rpath /usr/local/lib hash.lo global_data.lo io_buf.lo parse_regressor.lo parse_primi
tives.lo unique_sort.lo cache.lo rand48.lo simple_label.lo multiclass.lo oaa.lo multilabel_oaa.lo ect.lo autolink.lo binary.
lo lrq.lo cost_sensitive.lo multilabel.lo label_dictionary.lo csoaa.lo cb.lo cb_adf.lo cb_algs.lo search.lo search_meta.lo search_sequencetask.lo search_dep_parser.lo search_hooktask.lo search_multiclasstask.lo search_entityrelationtask.lo search_graph.lo parse_example.lo scorer.lo network.lo parse_args.lo accumulate.lo gd.lo learner.lo lda_core.lo gd_mf.lo mf.lo bfgs.lo noop.lo print.lo example.lo parser.lo loss_functions.lo sender.lo nn.lo bs.lo cbify.lo topk.lo stagewise_poly.lo log_multi.lo active.lo kernel_svm.lo best_constant.lo ftrl.lo svrg.lo lrqfa.lo liballreduce.la
libtool: link: g++  -fPIC -DPIC -shared -nostdlib /usr/lib/gcc/x86_64-linux-gnu/4.8/../../../x86_64-linux-gnu/crti.o /usr/lib/gcc/x86_64-linux-gnu/4.8/crtbeginS.o  .libs/hash.o .libs/global_data.o .libs/io_buf.o .libs/parse_regressor.o .libs/parse_primitives.o .libs/unique_sort.o .libs/cache.o .libs/rand48.o .libs/simple_label.o .libs/multiclass.o .libs/oaa.o .libs/multilabel_oaa.o .libs/ect.o .libs/autolink.o .libs/binary.o .libs/lrq.o .libs/cost_sensitive.o .libs/multilabel.o .libs/label_dictionary.o .libs/csoaa.o .libs/cb.o .libs/cb_adf.o .libs/cb_algs.o .libs/search.o .libs/search_meta.o .libs/search_sequencetask.o .libs/search_dep_parser.o .libs/search_hooktask.o .libs/search_multiclasstask.o .libs/search_entityrelationtask.o .libs/search_graph.o .libs/parse_example.o .libs/scorer.o .libs/network.o .libs/parse_args.o .libs/accumulate.o .libs/gd.o .libs/learner.o .libs/lda_core.o .libs/gd_mf.o .libs/mf.o .libs/bfgs.o .libs/noop.o .libs/print.o .libs/example.o .libs/parser.o .libs/loss_functions.o .libs/sender.o .libs/nn.o .libs/bs.o .libs/cbify.o .libs/topk.o .libs/stagewise_poly.o .libs/log_multi.o .libs/active.o .libs/kernel_svm.o .libs/best_constant.o .libs/ftrl.o .libs/svrg.o .libs/lrqfa.o   -Wl,-rpath -Wl,/home/marko/vw/2/vowpal_wabbit/vowpalwabbit/.libs -L/usr/local/lib -lboost_program_options -L/lib -lz -lpthread ./.libs/liballreduce.so -L/usr/lib/gcc/x86_64-linux-gnu/4.8 -L/usr/lib/gcc/x86_64-linux-gnu/4.8/../../../x86_64-linux-gnu -L/usr/lib/gcc/x86_64-linux-gnu/4.8/../../../../lib -L/lib/x86_64-linux-gnu -L/lib/../lib -L/usr/lib/x86_64-linux-gnu -L/usr/lib/../lib -L/usr/lib/gcc/x86_64-linux-gnu/4.8/../../.. -lstdc++ -lm -lc -lgcc_s /usr/lib/gcc/x86_64-linux-gnu/4.8/crtendS.o /usr/lib/gcc/x86_64-linux-gnu/4.8/../../../x86_64-linux-gnu/crtn.o  -O3   -Wl,-soname -Wl,libvw.so.0 -o .libs/libvw.so.0.0.0
.libs/cb_adf.o: In function `ec_is_label_definition(example&)':
cb_adf.cc:(.text+0x320): multiple definition of `ec_is_label_definition(example&)'
.libs/csoaa.o:csoaa.cc:(.text+0x560): first defined here
.libs/cb_adf.o: In function `ec_seq_is_label_definition(v_array<example*>)':
cb_adf.cc:(.text+0x3a0): multiple definition of `ec_seq_is_label_definition(v_array<example*>)'
.libs/csoaa.o:csoaa.cc:(.text+0x5e0): first defined here
.libs/cb_adf.o: In function `global_print_newline(vw&)':
cb_adf.cc:(.text+0x500): multiple definition of `global_print_newline(vw&)'
.libs/csoaa.o:csoaa.cc:(.text+0x1850): first defined here
.libs/cb_adf.o: In function `output_example(vw&, example&, bool&, v_array<example*>*)':
cb_adf.cc:(.text+0x650): multiple definition of `output_example(vw&, example&, bool&, v_array<example*>*)'
.libs/csoaa.o:csoaa.cc:(.text+0x19a0): first defined here
collect2: error: ld returned 1 exit status
make[2]: *** [libvw.la] Error 1
make[2]: Leaving directory `/home/marko/vw/2/vowpal_wabbit/vowpalwabbit'
make[1]: *** [all] Error 2
make[1]: Leaving directory `/home/marko/vw/2/vowpal_wabbit/vowpalwabbit'
make: *** [all-recursive] Error 1
```

**Comment by marko-asplund** [comment]
This happens on Ubuntu 14.04.2 (64-bit) and Boost v1.58.0.

**Comment by arielf** [comment]
Hi Marko,

You're making things more complex than they should be :)

You have a non standard setup.

Boost v1.58.0 is not standard on Ubuntu 14.04 (1.54 is).

`/usr/local` hints on more non standard stuff.  Normally on Ubuntu standard bost headers are installed under `/usr/include/boost` and the boost libraries are under `/usr/lib/x86_64-linux-gnu/` for example, the program options lib is at `/usr/lib/x86_64-linux-gnu/libboost_program_options.so`

How did you install boost 1.58 on Ubuntu?  From source?

There's almost never a need to build from source on Ubuntu.  Just install standard stuff, in standard locations, and things should work.

Try the following:

    sudo apt-get install libboost-dev

Now you have boost installed in the standard place.

Then start from a clean slate:

    git clone git://github.com/JohnLangford/vowpal_wabbit.git
    cd vowpal_wabbit
    make
    make test

That's it. Everything should be smooth and just work.

Summary:
  * Use the original/standard Makefile that comes with vowpal-wabbit.
  * Do not install boost from source
  * Do not run `autogen.sh` (it overwrites the original Makefile)

HTH

**Comment by arielf** [comment]
Just tried the above and found there was one minor issue:  The symlink to the libboost_program_options.so.<version> was missing. I fixed it with:

```
cd /usr/lib/x86_64-linux-gnu/
sudo ln -s libboost_program_options.so.1.54.0 libboost_program_options.so
```
Now the library is found by the linker.

**Comment by marko-asplund** [comment]
Hi Ariel,

Thanks for your reply.

Since the main README.md describes `autogen.sh` as a valid build method, so I decided to try that out.
If the use of `autogen.sh` build method is discouraged, I think it would make sense to mention that in README.md, or better yet, remove `autogen.sh` references entirely.

If Boost is recommended to be installed in a certain manner, or only certain Boost versions are supported, IMHO these details should be documented. I tried to find info on VW-specific Boost setup, but couldn't find any.

**Comment by JohnLangford** [comment]
This should be fixed in the master now.

-John

**Comment by arielf** [comment]
Hi Marko,

I wasn't very clear, sorry!.

What I meant to say is not that other versions of `boost` or `autogen.sh` are not supported.  They generally are (looks like in this case C++ namespaces got confused and John just fixed it).

All I wanted to say is that building dependent components from source tends to be more involved and error prone, and that on Ubuntu it is not needed.  IOW: on Ubuntu the simplest method (just `make`) is preferred. I will update REAME.md to say it.  Thanks.

HTH

**Comment by marko-asplund** [comment]
@JohnLangford Just tested building again and this issue seems to have gotten fixed by commit 0b1215ca6aac7147549e5e3d5ba6650b0684d233. Thanks!

@arielf Thanks for the background info and for updating the README.
